expo/eas-build
# [ENG-8775][build-tools][eas-build-job] add credentials to custom builds

# Why

https://linear.app/expo/issue/ENG-8775/implement-easprepare-credentials-and-easclean-up-credentials-ios

# How

Add the `sharedEasContext` object to the `BuildContext` class. It will be the object using which we will be passing some context information from `eas` namespaced functions to other steps.

Add new `eas/prepare_credentials` and `eas/clean_up_credentials`. Make sure that the `eas/clean_up_credentials` function is always executed (even if a previous step failed). These functions pass information to other steps using the `sharedEasContext` object.

Interpolate the `easCtx` object with commands and inputs to make it globally available from every step. Validate if referenced part of `easCtx` exists when accessing it (runtime). Validate in `BuildWorkflowValidator` if the selected property from the `easCtx` is even possible to be accessed.

While testing I was using changes from https://github.com/expo/eas-cli/pull/1859 to provide credentials to custom builds.

# Test Plan

1. Added automated tests
2. Run builds locally using
```yml
build:
  name: Run tests
  steps:
    - eas/checkout
    - eas/use_npm_token
    - eas/install_node_modules
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
    - eas/prepare_credentials
    - run:
        name: Check if Android credentials are available
        command: |
          echo "Android Keystore Path: ${ easCtx.credentials.android.keystore.keystorePath }"
    - eas/clean_up_credentials
```
workflow

iOS build:
![Screenshot 2023-06-01 at 14 47 45](https://github.com/expo/eas-build/assets/55145344/2a7ed2b7-c541-4213-8e72-ead6e61d3f47)

Android build:
![Screenshot 2023-06-01 at 15 51 57](https://github.com/expo/eas-build/assets/55145344/b93f8c7b-7637-4b85-9f3e-f5cc68a2a92b)

3. Run build locally using
```yml
build:
  name: Run tests
  steps:
    - eas/checkout
    - eas/use_npm_token
    - eas/install_node_modules
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
    - eas/prepare_credentials
    - run:
        name: Check if Android credentials are available
        command: |
          echo "${ easCtx.credentials.android.keystore.invalid }"
    - eas/clean_up_credentials
```
workflow
![Screenshot 2023-06-01 at 14 40 11](https://github.com/expo/eas-build/assets/55145344/d0c1cc14-d4b6-4b5f-a3a4-1ed63523a83e)

4. Run builds locally using
```yml
build:
  name: Run tests
  steps:
    - eas/checkout
    - eas/use_npm_token
    - eas/install_node_modules
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
    - eas/prepare_credentials
    - run:
        name: Check if iOS credentials are available
        command: |
          echo "iOS keystore path: ${ easCtx.credentials.ios.keychainPath }"
    - eas/clean_up_credentials
```
workflow

iOS build:
![Screenshot 2023-06-01 at 16 07 44](https://github.com/expo/eas-build/assets/55145344/1169dbda-5761-48ae-a804-903f115d154f)


Android build:
![Screenshot 2023-06-01 at 16 06 23](https://github.com/expo/eas-build/assets/55145344/21d53124-93c8-47eb-972b-1a91a670d4c1)

EDIT: I made the `eas/clean_up_credentials` function run even if some previous steps failed

![Screenshot 2023-06-01 at 16 44 25](https://github.com/expo/eas-build/assets/55145344/1d5abe50-2dda-430b-8947-19c61c0c5dfc)
